06_Userspace/04_System_Calls.md

```diff
@@ -123,13 +123,35 @@ In summary: if the kernel is on a 32-bit platform, use `sysenter`/`sysexit`, for
 
 ### Using Syscall & Sysret
 
-Before using these instructions we'll need to perform a bit of setup first. They require the GDT to have a very specific layout.
+Before using these instructions we'll need to perform a bit of setup. 
 
-Let's assume that our kernel CS is 0x8: to use `syscall` the kernel SS **must** be the next GDT entry, at offset 0x10.
+First things first, the `syscall`/`sysret` pair uses three MSRs: `STAR` (0xC0000081), `LSTAR` (0xC0000082) and `FMASK` (0xC0000084).
 
-For `sysret`, which returns to user mode, things are a little more complex. This instruction allows for going to both compatibility mode (32-bit long mode) and long mode (64-bit long mode proper). So the instruction actually requires three GDT selectors to be placed immedately following each other: user CS (32-bit mode), user SS, user CS (64-bit mode). As an example if our 32-bit CS was 0x18, our user SS **must** be at 0x20, and our 64-bit user CS **must** be at 0x28.
+`STAR` has three fields used to set the GDT descriptors of userspace and kernelspace. The low 32 bits are reserved for the handler of 32-bit variant of `syscall`. It's safe to set it to zero if you don't target 64-bit code. The next 16 bits (47:32) hold the kernel CS/SS base. When `syscall` is executed, the new CS is set to `base` and the new SS is set to `base + 8`. The last 16 bits (64:48) hold the userspace CS/SS base. Because it supports jumping back to compatibility mode, `sysret` requires a very specific GDT layout. It expects the 32-bit CS to be at `base`, the SS (either 64-bit or 32-bit) at `base + 8` and the 64-bit CS to be at `base + 16`. If you don't target 32-bit code, it's safe to omit the 32-bit segment from the GDT, but not that the 64-bit CS will still be fetched from `base + 16`.
```

Comment:
> used to set GDT descriptors of

maybe "that specify the selectors used for", a bit more accurate.

> if you dont target 64-bit code

I think this is backwards, you mean its safe to set to zero if you dont support a 32-bit userspace?

> but not that the 64-bit CS

I think that should be "but note that the 64-bit CS"